test/integration/test_recalculate_user_disk_usage.py

```diff
@@ -0,0 +1,39 @@
+from galaxy_test.base.populators import DatasetPopulator
+from galaxy_test.driver.integration_util import IntegrationTestCase
+
+
+class TestRecalculateUserDiskUsageIntegration(IntegrationTestCase):
+    task_based = True
+    dataset_populator: DatasetPopulator
+
+    def setUp(self):
+        super().setUp()
+        self.dataset_populator = DatasetPopulator(self.galaxy_interactor)
+
+    @classmethod
+    def handle_galaxy_config_kwds(cls, config):
+        super().handle_galaxy_config_kwds(config)
+        config["allow_user_dataset_purge"] = True
+
+    def test_recalculate_user_disk_usage(self):
+        # The initial disk usage is 0
+        current_usage = self.dataset_populator.get_usage_for(None)
+        assert current_usage["total_disk_usage"] == 0
+        size = 100
+        history_id = self.dataset_populator.new_history()
+        hda_id = self.dataset_populator.new_dataset(history_id, content=f"{'0'*size}", wait=True)["id"]
+        expected_usage = size + 1  # +1 for the new line character in the dataset
+        # The usage should be the total of the datasets
+        current_usage = self.dataset_populator.get_usage_for(None)
+        assert current_usage["total_disk_usage"] == expected_usage
+        self.dataset_populator.delete_dataset(history_id, hda_id, purge=True, wait_for_purge=True)
+        # After purging the disk usage might not be automatically recalculated yet
+        current_usage = self.dataset_populator.get_usage_for(None)
+
```

Comment:
So this isn't updated and this led me to realizing there is a quota adjustment that isn't flushed down this path of the API (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/managers/hdas.py#L219). My guess is this was fine when I implemented years ago because we were flushing every session then - Celery and async means we need an explicit flush there.

Fixing that code to flush fixes this test case but... it only fixes the test case because then ``recalculate_disk_usage`` doesn't need to do any work. I'll to figure out that bug with the bug that makes things get un-synced still in 😰.

Reply:
Fixing the bug that let the test case work (purging not updating the quota) and fixing the bug of recalculating not working (the SA attribute was somehow being diff'ed with changes in the database and updating against the expected value causing a negative value (I think) regardless of the source though expiring disk_usage fixes it).

The problem is that we're not really testing recalculation - I think that needs to be a unit test right? Anytime you can cause that to be off via the API - we've found a bug that we should fix such as this here.

Reply:
Thank you so much for figuring it out!

So, if I understood it correctly, the `recalculate_disk_usage` API endpoint is not really needed anymore? or is there still some possible desync that calling this recalculate_disk_usage will fix?

Reply:
The disk usage stuff would become un-synced before we added the celery purging. I think there are other bugs in there with tracking - I don't know where they are and I think the API endpoint is still needed at a result. I am not sure the best way to test it. One idea is to recreate the problem with the model tests below in test_quota.py and verify recalculate works with unit tests. Luckily the celery task is a very small wrapper around the model code so that feels like a pretty valuable test. I'll add that test really quick here and then approve this PR.

Reply:
I've added test_calculate_usage_readjusts_incorrect_quota below.